src/rust/cryptography-x509-validation/src/lib.rs

```diff
@@ -11,6 +11,349 @@ pub mod policy;
+    Other(String),
+}
+
+impl From<asn1::ParseError> for ValidationError {
+    fn from(value: asn1::ParseError) -> Self {
+        Self::Malformed(value)
+    }
+}
+
+impl From<DuplicateExtensionsError> for ValidationError {
+    fn from(value: DuplicateExtensionsError) -> Self {
+        Self::DuplicateExtension(value)
+    }
+}
+
+#[derive(Default)]
+struct AccumulatedNameConstraints<'a> {
+    sans: Vec<GeneralName<'a>>,
+    name_constraints: Vec<NameConstraints<'a>>,
+}
+
+impl<'a> AccumulatedNameConstraints<'a> {
+    fn accumulate_san(&mut self, extensions: &Extensions<'a>) -> Result<(), ValidationError> {
+        if let Some(sans) = extensions.get_extension(&SUBJECT_ALTERNATIVE_NAME_OID) {
+            let sans: SubjectAlternativeName<'_> = sans.value()?;
+            self.sans.extend(sans);
+        }
+
+        Ok(())
+    }
+
+    fn apply_inner(
+        &self,
+        constraint: &GeneralName<'a>,
+        san: &GeneralName<'_>,
+    ) -> Result<ApplyNameConstraintStatus, ValidationError> {
+        match (constraint, san) {
+            (GeneralName::DNSName(pattern), GeneralName::DNSName(name)) => {
+                match (DNSConstraint::new(pattern.0), DNSName::new(name.0)) {
+                    (Some(pattern), Some(name)) => Ok(Applied(pattern.matches(&name))),
+                    (Some(_), None) => Err(ValidationError::Other(format!(
+                        "unsatisfiable DNS name constraint: NC {} cannot match SAN {}",
+                        pattern.0, name.0
+                    ))),
+                    (None, _) => Err(ValidationError::Other(format!(
+                        "malformed DNS name constraint: {}",
+                        pattern.0
+                    ))),
+                }
+            }
+            (GeneralName::IPAddress(pattern), GeneralName::IPAddress(name)) => {
+                match (
+                    IPConstraint::from_bytes(pattern),
+                    IPAddress::from_bytes(name),
+                ) {
+                    (Some(pattern), Some(name)) => Ok(Applied(pattern.matches(&name))),
+                    (Some(_), None) => Err(ValidationError::Other(format!(
+                        "unsatisfiable IP name constraint: NC {:?} cannot match SAN {:?}",
+                        pattern, name,
+                    ))),
+                    (None, _) => Err(ValidationError::Other(format!(
+                        "malformed IP name constraints: {:?}",
+                        pattern
+                    ))),
+                }
+            }
+            _ => Ok(Skipped),
+        }
+    }
+
+    /// Apply the current name constraints (including those in the specified
+    /// extensions, if any) to the accumulated SAN set.
+    ///
+    /// On success (no constraint violations found), the new constraints
+    /// are additionally added to the name constraint set for future checks.
+    fn apply_and_accumulate(&mut self, extensions: &Extensions<'a>) -> Result<(), ValidationError> {
```

Comment:
This should probably be `apply_and_accumulate_name_constraints` otherwise it sounds like it's the same accumulation as `acculumate_san`